NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,532 @@
+	std::string content = EncodeJSON<context>(sqvm);
+	if (ContainsInvalidChars(content))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm, fmt::format("File contents may not contain NUL/\\0 characters! Make sure your strings are valid!", mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	fs::create_directories(dir);
+	// this actually allows mods to go over the limit, but not by much
+	// the limit is to prevent mods from taking gigabytes of space,
+	// this ain't a cloud service.
+	if (GetSizeOfFolderContentsMinusFile(dir, fileName) + content.length() > MAX_FOLDER_SIZE)
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"The mod {} has reached the maximum folder size.\n\nAsk the mod developer to optimize their data usage,"
+				"or increase the maximum folder size using the -maxfoldersize launch parameter.",
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	g_pSaveFileManager->SaveFileAsync<context>(dir / fileName, content);
+
+	return SQRESULT_NULL;
+}
+
+// int NS_InternalLoadFile(string file)
+ADD_SQFUNC("int", NS_InternalLoadFile, "string file", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm, 1); // the function that called NSLoadFile :)
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	fs::path dir = savePath / fs::path(mod->m_ModDirectory).filename();
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (!IsPathSafe(fileName, dir))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"File name invalid ({})! Make sure it does not contain any non-ASCII character, and results in a path inside your mod's "
+				"save folder.",
+				fileName,
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	g_pSquirrel<context>->pushinteger(sqvm, g_pSaveFileManager->LoadFileAsync<context>(dir / fileName));
+
+	return SQRESULT_NOTNULL;
+}
+
+// bool NSDoesFileExist(string file)
+ADD_SQFUNC("bool", NSDoesFileExist, "string file", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm);
+
+	fs::path dir = savePath / fs::path(mod->m_ModDirectory).filename();
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (!IsPathSafe(fileName, dir))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"File name invalid ({})! Make sure it does not contain any non-ASCII character, and results in a path inside your mod's "
+				"save folder.",
+				fileName,
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	g_pSquirrel<context>->pushbool(sqvm, fs::exists(dir / (fileName)));
+	return SQRESULT_NOTNULL;
+}
+
+// int NSGetFileSize(string file)
+ADD_SQFUNC("int", NSGetFileSize, "string file", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
```

Comment:
Same thing.

Reply:
Fair point, what would possible usecases for this be?

Reply:
Checking how much space the mod has left to store things?

Reply:
fair

Reply:
Tbh I really don't foresee any mods actually going over the limit, except if someone does some like logging system in script or something weird

Reply:
> Checking how much space the mod has left to store things?

Could maybe add a function for that specifically instead?

Reply:
> Could maybe add a function for that specifically instead?

I think that this is probably a good idea, The only use case (that is sane) that I can think of is to check how much space is remaining, but that would just mean the modder has to do `MAX_FILE_SIZE_CONST_THING - NSGetFileSize()` to actually get the value that they want.

Speaking of having to use a const, the size limitation const should imo be exposed to squirrel so that modders can ensure that they arent trying to write excessively sized files or something

Reply:
Will add in next commit `NSGetModSizeLimit` to get the size limit of a save folder, and `NSGetTotalSpaceRemaining` which returns the space remaining for a save folder until it reaches the cap.

Reply:
> Will add in next commit `NSGetModSizeLimit` to get the size limit of a save folder

Would it be better to push the const used here to the squirrel VM directly instead of making a function for it?

Reply:
True, will call it `MAX_FOLDER_SIZE`.

Reply:
Resolved in https://github.com/R2Northstar/NorthstarLauncher/commit/df0554a0e7cd8995eb8ecfaf045f659f99785b34